### PR TITLE
CreateRoom permission for Sign-in + protecting backend

### PR DESCRIPTION
**Pull request by hadicheaito1**
## Description

- Back-end endpoints are protected if user does not have permission
- User is redirected to `MainPage` on sign in if they don't have proper permission
## Testing Steps

- Turn `CreateRoom` permission to false for User role 
- Log in as regular user role
- Should be redirected to `/main-page` instead of `rooms`
## TODO
- Do the same for sign in
- UI

### CODE REVIEW

**Comment by hadicheaito1** [comment]
Suggestions for this new page/route name? Something instead of `/main-page` ?

File: app/controllers/api/v1/rooms_controller.rb [hunk 1]
```diff
@@ -6,11 +6,12 @@ class RoomsController < ApiController
       skip_before_action :ensure_authenticated, only: %i[public_show]
 
       before_action :find_room, only: %i[show update destroy recordings recordings_processing purge_presentation public_show]
-
+      before_action except: %i[public_show] do
```
Comment (hadicheaito1) [inline app/controllers/api/v1/rooms_controller.rb:9]: I added the `except` after looking at the rspec test.
I don't fully understand why the `session[:user_id] = nil` in the rspec test for this. Could someone explain the usage of `public_show`

**Comment by scouillard** [comment]
Shouldn't the "createRoom" permission just deal with the "create room" button pretty much? 
A user without the permission would still be able to belong to SharedRooms, and access /rooms, right?

File: app/controllers/api/v1/rooms_controller.rb (hunk 1 above)
Comment (scouillard) [reply app/controllers/api/v1/rooms_controller.rb:9]: `session[:user_id] = nil` is to log off the current_user. So in specs it is usually to make sure some one not logged in does not have the access to something.

**Comment by hadicheaito1** [comment]
> Shouldn't the "createRoom" permission just deal with the "create room" button visibility essentially? A user without the permission would still be able to belong to SharedRooms, and access /rooms, right?

I wish :sweat_smile: 
From what I understand, if `CreateRoom` permission is set to false, then it disallows the user from creating rooms, accessing any room (as well as SharedRooms and all that), as well as recordings (since recordings belong to rooms).

The permission name is misleading imo

File: app/controllers/api/v1/rooms_controller.rb (hunk 1 above)
Comment (hadicheaito1) [reply app/controllers/api/v1/rooms_controller.rb:9]: Yeah i'm aware that it signs out the user. I'm more looking for why do that for this specific test and why do they still have access to `public_show` even thought they are not logged in

Comment (scouillard) [reply app/controllers/api/v1/rooms_controller.rb:9]: Right. I'm also unsure of what is going on with `public_show` (and `publicRoom` from #3836)

**Comment by scouillard** [comment]
Wow, okay.
Maybe we should redirect those users to our landing page at '/' and maybe remove the sign in sign up button if the user is signed in?
We also have to figure out something for '/' because right now it crashes if a logged in user tries to access it.

**Comment by hadicheaito1** [comment]
> Wow, okay. Maybe we should redirect those users to our landing page at '/' and maybe remove the sign in sign up button if the user is signed in? I mean it seems like there is not much those users can do but have access to their profile via the navbar. We also have to figure out something for '/' because right now it crashes if a logged in user tries to access it.

Based on v2, the users without this permission are able to access their own profile and are able to paste a join url to a meeting. This might change for v3.

Yeah it's crashing because of the forbidden status being returned. Soon we are going to have to make a page for this case so that something is rendered and it doesn't just break the app

**Comment by scouillard** [comment]
I think that is how I would proceed for the '/' homepage.
Case 1 - User is not logged
Regular homepage, full height without navbar.
Case 2 - User is logged in && has createRoom permission
'/' redirects to '/rooms'
Case 3 - User is logged in && does NOT have createRoom permission
Regular homepage with navbar.

**Comment by scouillard** [comment]
> > Wow, okay. Maybe we should redirect those users to our landing page at '/' and maybe remove the sign in sign up button if the user is signed in? I mean it seems like there is not much those users can do but have access to their profile via the navbar. We also have to figure out something for '/' because right now it crashes if a logged in user tries to access it.
> 
> Based on v2, the users without this permission are able to access their own profile and are able to paste a join url to a meeting. This might change for v3.
> 
> Yeah it's crashing because of the forbidden status being returned. Soon we are going to have to make a page for this case so that something is rendered and it doesn't just break the app

I'm sorry I might be dragging out of this PR scope - what is returning forbidden?

File: spec/controllers/rooms_controller_spec.rb [hunk 2]
```diff
@@ -3,9 +3,9 @@
 require 'rails_helper'
 
 RSpec.describe Api::V1::RoomsController, type: :controller do
-  let(:user) { create(:user) }
-  let(:user_with_manage_rooms_permission) { create(:user, :with_manage_rooms_permission) }
-  let(:user_with_manage_users_permission) { create(:user, :with_manage_users_permission) }
+  let(:user) { create(:user, :with_create_room_permission) }
+  let(:user_with_manage_rooms_permission_and_create_room_permission) { create(:user, :with_manage_rooms_permission, :with_create_room_permission) }
+  let(:user_with_manage_users_permission_and_create_room_permission) { create(:user, :with_manage_users_permission, :with_create_room_permission) }
```
Comment (scouillard) [inline spec/controllers/rooms_controller_spec.rb:8]: Those are duplicated

**Comment by hadicheaito1** [comment]
> > > Wow, okay. Maybe we should redirect those users to our landing page at '/' and maybe remove the sign in sign up button if the user is signed in? I mean it seems like there is not much those users can do but have access to their profile via the navbar. We also have to figure out something for '/' because right now it crashes if a logged in user tries to access it.
> > 
> > 
> > Based on v2, the users without this permission are able to access their own profile and are able to paste a join url to a meeting. This might change for v3.
> > Yeah it's crashing because of the forbidden status being returned. Soon we are going to have to make a page for this case so that something is rendered and it doesn't just break the app
> 
> I'm sorry I might be dragging out of this PR scope - what is returning forbidden? The crash on '/' if logged in is in V3 current state.

My bad I misread your comment. Ignore the forbidden status stuff

File: spec/controllers/rooms_controller_spec.rb (hunk 2 above)
Comment (hadicheaito1) [reply spec/controllers/rooms_controller_spec.rb:8]: One is for users and one is for rooms

File: spec/controllers/recordings_controller_spec.rb [hunk 3]
```diff
@@ -3,8 +3,10 @@
 require 'rails_helper'
 
 RSpec.describe Api::V1::RecordingsController, type: :controller do
-  let(:user) { create(:user) }
-  let(:user_with_manage_recordings_permission) { create(:user, :with_manage_recordings_permission) }
+  let(:user) { create(:user, :with_create_room_permission) }
+  let(:user_with_manage_recordings_permission_and_create_room_permission) do
+    create(:user, :with_manage_recordings_permission, :with_create_room_permission)
+  end
```
Comment (scouillard) [inline spec/controllers/recordings_controller_spec.rb:9]: The createRoom permission is so restrictive, maybe we should just enable it by default with every other permissions?
Else we will have to do this for all of our specs pretty much, no?
Maybe we could have a negative logic trait `:without_create_room_permission` instead and use it when we need to test the createRoom permission.

File: spec/controllers/recordings_controller_spec.rb [hunk 4]
```diff
@@ -3,8 +3,10 @@
 require 'rails_helper'
 
 RSpec.describe Api::V1::RecordingsController, type: :controller do
-  let(:user) { create(:user) }
-  let(:user_with_manage_recordings_permission) { create(:user, :with_manage_recordings_permission) }
+  let(:user) { create(:user, :with_create_room_permission) }
```
Comment (scouillard) [inline spec/controllers/recordings_controller_spec.rb:6]: I dont think the :with_create_room_permission trait belongs here

File: spec/controllers/recordings_controller_spec.rb (hunk 3 above)
Comment (scouillard) [reply spec/controllers/recordings_controller_spec.rb:9]: Something like:
```
  let(:user_with_manage_recordings_permission) do
    create(:user, :with_manage_recordings_permission, :with_create_room_permission)
  end
```

And if needed:
```
  let(:user_with_manage_recordings_permission) do
    create(:user, :with_manage_recordings_permission, :without_create_room_permission)
  end
```

File: spec/controllers/rooms_controller_spec.rb (hunk 2 above)
Comment (scouillard) [reply spec/controllers/rooms_controller_spec.rb:8]: Lol right I'm sorry

File: spec/controllers/recordings_controller_spec.rb (hunk 3 above)
Comment (scouillard) [reply spec/controllers/recordings_controller_spec.rb:9]: Actually, maybe there is a way to treat a user without the createRoom permission almost like an unauthorised user instead?

File: app/javascript/hooks/mutations/sessions/useCreateSession.jsx [hunk 5]
```diff
@@ -6,13 +6,20 @@ import axios from '../../../helpers/Axios';
...
     {
       onSuccess: () => {
         queryClient.invalidateQueries('useSessions');
-        navigate('/rooms');
+        // if the current user does NOT have the CreateRoom permission, then do not re-direct to rooms page
+        if (createRoomPermission === 'false') {
+          navigate('/main-page');
+        } else {
+          navigate('/rooms');
+        }
```
Comment (farhatahmad) [inline app/javascript/hooks/mutations/sessions/useCreateSession.jsx:22]: This is kind of weird. Can we not do all of this logic in the `onSuccess` if you just add an argument to it? `onSuccess: (data) ...`

File: app/javascript/main.jsx [hunk 6]
```diff
@@ -52,6 +53,7 @@ const root = (
             <Route path="/rooms" element={<Rooms />} />
             <Route path="/rooms/:friendlyId" element={<Room />} />
             <Route path="/rooms/:friendlyId/join" element={<RoomJoin />} />
+            <Route path="/main-page" element={<MainPage />} />
```
Comment (farhatahmad) [inline app/javascript/main.jsx:56]: I would call it `/home`

File: spec/controllers/recordings_controller_spec.rb (hunk 3 above)
Comment (farhatahmad) [reply spec/controllers/recordings_controller_spec.rb:9]: Yeah, by default users should have `create room` enabled. We can do it in an `after_create` in the factory

**Review by farhatahmad — APPROVED** [review]

**Pull request merged**